build: Propagate user-defined tools to native packages

hebasto opened the pull request:
On master (66636ca438cb65fb18bcaa4540856cef0cee2029) build tools for native packages are set to [defaults](https://github.com/bitcoin/bitcoin/blob/66636ca438cb65fb18bcaa4540856cef0cee2029/depends/builders/default.mk#L1-L9), even when such a tool is not available in the system by its default name:
```
$ cd depends
$ make print-native_capnp_cxx MULTIPROCESS=1 CC=clang CXX=clang++
native_capnp_cxx=g++
```

On system without GCC installed, this causes an error:
```
$ make native_capnp_configured MULTIPROCESS=1 CC=clang CXX=clang++
Configuring native_capnp...
...
configure: error: *** A compiler with support for C++14 language features is required.
make: *** [funcs.mk:283: /home/hebasto/GitHub/bitcoin/depends/work/build/x86_64-pc-linux-gnu/native_capnp/0.7.0-4f60a88ce07/./.stamp_configured] Error 1
```

This PR fixes this issue with introducing support of `{CC,CXX}_FOR_BUILD` variables.

With this PR:
```
$ make print-native_capnp_cxx MULTIPROCESS=1 CC_FOR_BUILD=clang CXX_FOR_BUILD=clang++
native_capnp_cxx=clang++
$ make native_capnp_configured MULTIPROCESS=1 CC_FOR_BUILD=clang CXX_FOR_BUILD=clang++
# no errors
```

hebasto committed:
build: Propagate user-defined tools to native packages

dongcarl commented:
Nice, perhaps we can use variable names like `CC_FOR_BUILD` to distinguish? This is what [autoconf](https://www.gnu.org/software/autoconf-archive/ax_prog_cc_for_build.html) uses, and is used in secp256k1: https://github.com/bitcoin-core/secp256k1/blob/793ad9016a09c3bf5a5f280c812c46526250d839/configure.ac#L342

hebasto commented:
> Concept ACK, perhaps we can use variable names like `CC_FOR_BUILD` to distinguish? This is what [autoconf](https://www.gnu.org/software/autoconf-archive/ax_prog_cc_for_build.html) uses, and is used in secp256k1: https://github.com/bitcoin-core/secp256k1/blob/793ad9016a09c3bf5a5f280c812c46526250d839/configure.ac#L342

It makes sense within a build system wich is based on Autotools. Is there any benefit in pure `Makefile`'s (our depends build system)?

dongcarl commented:
The benefit is just that the user can specify a different CC for build vs target, e.g.:
```sh
$ make -C depends CC_FOR_BUILD=clang CC=x86_64-w64-mingw32-gcc
```

hebasto commented:
> The benefit is just that the user can specify a different CC for build vs target, e.g.:
> 
> ```shell
> $ make -C depends CC_FOR_BUILD=clang CC=x86_64-w64-mingw32-gcc
> ```

Should new variables be limited to compilers only, i.e, `CC_FOR_BUILD` and `CXX_FOR_BUILD`, or cover the whole tool set?

dongcarl commented:
> Should new variables be limited to compilers only, i.e, `CC_FOR_BUILD` and `CXX_FOR_BUILD`, or cover the whole tool set?

Up to you! It seems like CC and CXX would be useful enough, so whatever's easy to do I guess

shaavan reviewed (commented):
Concept ACK

This PR allows users to provide their values for CC and CXX flags. If the user does not give those values, the earlier method of assigning the build_$(build_os)_$1, and build_$(build_arch)_$(build_os)_$1.
Otherwise, these variables are assigned the value the user has provided. If that is empty, the earlier set value of these variables is used. If that's also empty, the default value is used. 

As discussed in the PR description, this change will prevent unnecessary problems where the default tools are not available (by the same name) in the system.

P.s.: (As it's my first time reviewing PR involving makefiles, if some of my observations are erroneous, please do correct me.)

hebasto committed:
build: Use `_FOR_BUILD` suffix for tools to build native packages

hebasto commented:
Updated 557d4d5d74219b38d0c1a37b7e30122db3e575a8 -> 0ac8f23a040a997aa545c054b7806a2a98aaca45 ([pr23571.01](https://github.com/hebasto/bitcoin/commits/pr23571.01) -> [pr23571.02](https://github.com/hebasto/bitcoin/commits/pr23571.02)).

Added a commit which address  @dongcarl's [comment](https://github.com/bitcoin/bitcoin/pull/23571#issuecomment-975774282):



> Concept ACK, perhaps we can use variable names like `CC_FOR_BUILD` to distinguish? This is what [autoconf](https://www.gnu.org/software/autoconf-archive/ax_prog_cc_for_build.html) uses, and is used in secp256k1: https://github.com/bitcoin-core/secp256k1/blob/793ad9016a09c3bf5a5f280c812c46526250d839/configure.ac#L342



> The benefit is just that the user can specify a different CC for build vs target, e.g.:
> 
> ```shell
> $ make -C depends CC_FOR_BUILD=clang CC=x86_64-w64-mingw32-gcc
> ```

dongcarl commented on depends/builders/default.mk:
```diff
@@ -8,9 +8,9 @@ default_build_OTOOL = otool
-ifeq ($(filter $(origin $1),undefined default),)
-build_$(build_os)_$1 = $(or $($1),$(build_$(build_os)_$1),$(default_build_$1))
-build_$(build_arch)_$(build_os)_$1 = $(or $($1),$(build_$(build_arch)_$(build_os)_$1),$$(build_$(build_os)_$1))
+ifneq ($(origin $1_FOR_BUILD),undefined)
+build_$(build_os)_$1 = $(or $($1_FOR_BUILD),$(build_$(build_os)_$1),$(default_build_$1))
+build_$(build_arch)_$(build_os)_$1 = $(or $($1_FOR_BUILD),$(build_$(build_arch)_$(build_os)_$1),$$(build_$(build_os)_$1))
```
Perhaps we should establish a style for Makefiles... But I find it more explicit to say `$($(1)_FOR_BUILD)` like we do for `$($(1)_postprocessed)` etc.

hebasto replied:
Sorry Carl, but I'm not sure if I understand you correctly...

If you mean s/`$($1_FOR_BUILD)`/`$($(1)_FOR_BUILD)`/, both are not equivalent though.

dongcarl replied:
Oh, then in that case shouldn't we be using `$($(1)_FOR_BUILD)` in these cases for correctness' sake?

hebasto replied:
> Oh, then in that case shouldn't we be using `$($(1)_FOR_BUILD)` in these cases for correctness' sake?

We are interested in `$(CXX_FOR_BUILD)` rather `$(g++_FOR_BUILD)`, no?

hebasto replied:
I was definitely wrong. Sorry for noise.

[Updated](https://github.com/bitcoin/bitcoin/pull/23571#issuecomment-998360080).

laanwj commented:
Concept ACK, I've run into this problem in the past too.

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
eeef2f1cf459b1ed2901b8aad5037f1af535de30b75a37e3bc8ce4b1b630c49e  guix-build-0ac8f23a040a/output/aarch64-linux-gnu/SHA256SUMS.part
5d8fd9e11874269c6e3a90cd644c76d6bc9723e3dc9d42ed10ec5d97eaac5fb1  guix-build-0ac8f23a040a/output/aarch64-linux-gnu/bitcoin-0ac8f23a040a-aarch64-linux-gnu-debug.tar.gz
5734ea5f9a29e0e67d97a457495e3ebb882688c5922e4f00bb0b87e0d1d3f3e8  guix-build-0ac8f23a040a/output/aarch64-linux-gnu/bitcoin-0ac8f23a040a-aarch64-linux-gnu.tar.gz
5be326c09d7ccf61381b546671952e5ec1f3d023ecfb72c8faa2d045b66446fe  guix-build-0ac8f23a040a/output/arm-linux-gnueabihf/SHA256SUMS.part
2d98bf2d15aa686978023522fb67cbb5afc773070e774e54ada7d4849ea469d3  guix-build-0ac8f23a040a/output/arm-linux-gnueabihf/bitcoin-0ac8f23a040a-arm-linux-gnueabihf-debug.tar.gz
5dacf0c4e62b685140d9d0aae5ffaa696917c5fb520f1d9204b0b8bd8cea6f3d  guix-build-0ac8f23a040a/output/arm-linux-gnueabihf/bitcoin-0ac8f23a040a-arm-linux-gnueabihf.tar.gz
8a308b43c16af5847dc50d8234abb80063e0a3cc5f1556f8b5e53f05316de474  guix-build-0ac8f23a040a/output/dist-archive/bitcoin-0ac8f23a040a.tar.gz
9bb864663c3f786ad5ff7ad99c258938a68e46d814134644c8c1bf2703dd9abf  guix-build-0ac8f23a040a/output/powerpc64-linux-gnu/SHA256SUMS.part
b7bd2a8d3d33ff330da075792389712db2743f0661b2d4870534b03b2fa21ec6  guix-build-0ac8f23a040a/output/powerpc64-linux-gnu/bitcoin-0ac8f23a040a-powerpc64-linux-gnu-debug.tar.gz
08cbd743f83dbcff02b8324eeb743619f8e7dc4eef61daf0f42ab0fba686ed19  guix-build-0ac8f23a040a/output/powerpc64-linux-gnu/bitcoin-0ac8f23a040a-powerpc64-linux-gnu.tar.gz
5b4ce19be8690cb9b8d9d675aa76d8bd540b2f712d01d7d1e661988c01fc4857  guix-build-0ac8f23a040a/output/powerpc64le-linux-gnu/SHA256SUMS.part
85f2ea989a6223f5f20a548b432bd3fd218f651bc3707f7a30d7310147d89fee  guix-build-0ac8f23a040a/output/powerpc64le-linux-gnu/bitcoin-0ac8f23a040a-powerpc64le-linux-gnu-debug.tar.gz
8e1ad04f2437f739f3243f17c975846b682ccfd33be13e442654f89fff7c2faa  guix-build-0ac8f23a040a/output/powerpc64le-linux-gnu/bitcoin-0ac8f23a040a-powerpc64le-linux-gnu.tar.gz
baba8b47ae464659805fe8f799b90ad921b74114db81c68ab16f0c59cc2528e1  guix-build-0ac8f23a040a/output/riscv64-linux-gnu/SHA256SUMS.part
00f500e08094e773347b38faba5147265b7bc106a90bfb47a02e870b48c66a1d  guix-build-0ac8f23a040a/output/riscv64-linux-gnu/bitcoin-0ac8f23a040a-riscv64-linux-gnu-debug.tar.gz
e5e6d2cb2f7eb96aff42ab471f936e241be83d2442a1ea6e31b3a276e67f3edc  guix-build-0ac8f23a040a/output/riscv64-linux-gnu/bitcoin-0ac8f23a040a-riscv64-linux-gnu.tar.gz
49792292ae4adca028f4d1acaf25c614687740cb03dd58b4b9c41d4b47bcf10a  guix-build-0ac8f23a040a/output/x86_64-apple-darwin19/SHA256SUMS.part
af95ed0b8ebd5f8879637a52848bd3ac57a44deee3bf15c258fedeb548dcc105  guix-build-0ac8f23a040a/output/x86_64-apple-darwin19/bitcoin-0ac8f23a040a-osx-unsigned.dmg
40287d1e76212247df6b365120ef97662c5f93c043a8348b2e854dd783a99e0b  guix-build-0ac8f23a040a/output/x86_64-apple-darwin19/bitcoin-0ac8f23a040a-osx-unsigned.tar.gz
97bc510a7dcc29fc6b43331f23e41f41ae5a4ee3faf25240af3da62ab5b74fee  guix-build-0ac8f23a040a/output/x86_64-apple-darwin19/bitcoin-0ac8f23a040a-osx64.tar.gz
631c5de077daa48b96ef75c5f85d116e9a6f3cb49dd1de25bf0a1c75098f347f  guix-build-0ac8f23a040a/output/x86_64-linux-gnu/SHA256SUMS.part
df87812daa747f878238ddc2b6580821dafc417b5889ed5d54d04ab11a6bab89  guix-build-0ac8f23a040a/output/x86_64-linux-gnu/bitcoin-0ac8f23a040a-x86_64-linux-gnu-debug.tar.gz
2c9e5ede94a7faa44f7df0c139a82a21d77788d6e3b097cf1888c45e262b67c1  guix-build-0ac8f23a040a/output/x86_64-linux-gnu/bitcoin-0ac8f23a040a-x86_64-linux-gnu.tar.gz
93dfd01edc575d559a111d2b91d49d96c6fee8433780a8deeee774d76a8d5622  guix-build-0ac8f23a040a/output/x86_64-w64-mingw32/SHA256SUMS.part
ea2347512b6e1d1b52135df27501f14f453c9a3fe2b8efb60e36e0bfe1628012  guix-build-0ac8f23a040a/output/x86_64-w64-mingw32/bitcoin-0ac8f23a040a-win-unsigned.tar.gz
30e143700f30eb12946d8c28b33c2886f7c2659fbeaafa19d195346dbf8a2f99  guix-build-0ac8f23a040a/output/x86_64-w64-mingw32/bitcoin-0ac8f23a040a-win64-debug.zip
9c4d57f2e9f1eee37cb6aff2c5f2903bb6ee08a03956e3dc68cc1b122546313a  guix-build-0ac8f23a040a/output/x86_64-w64-mingw32/bitcoin-0ac8f23a040a-win64-setup-unsigned.exe
cc7883e5cb48b5d30d027e2b007e983744874b6c6951a9b3a3fade8176d6a04a  guix-build-0ac8f23a040a/output/x86_64-w64-mingw32/bitcoin-0ac8f23a040a-win64.zip
```

hebasto commented:
Updated 0ac8f23a040a997aa545c054b7806a2a98aaca45 -> 28725337e6011789a6b4e80c0cb127979bf34221 ([pr23571.02](https://github.com/hebasto/bitcoin/commits/pr23571.02) -> [pr23571.03](https://github.com/hebasto/bitcoin/commits/pr23571.03)).

Addressed  @dongcarl's [comment](https://github.com/bitcoin/bitcoin/pull/23571#discussion_r759517867).

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
a33d7a344036242596b631a5601cfce1c1dee5e76581d97cce351a2a772f5c66  guix-build-28725337e601/output/aarch64-linux-gnu/SHA256SUMS.part
3d88200cb088296a52fcb40a167aa483de2012d9021eddc8e754d1d14f5a1b50  guix-build-28725337e601/output/aarch64-linux-gnu/bitcoin-28725337e601-aarch64-linux-gnu-debug.tar.gz
65fac30ac6f5e495484c35d8d2d531806a0ccb432bf59f0e08f3d5c954434b64  guix-build-28725337e601/output/aarch64-linux-gnu/bitcoin-28725337e601-aarch64-linux-gnu.tar.gz
ff89320d0db9be6b55b7ddf611964fa9adb81311460c754c934d7141984b5575  guix-build-28725337e601/output/arm-linux-gnueabihf/SHA256SUMS.part
b29142d2ff4ebd0897a54a47568b470f0e5b39dd634b643325483eabc6624972  guix-build-28725337e601/output/arm-linux-gnueabihf/bitcoin-28725337e601-arm-linux-gnueabihf-debug.tar.gz
70d98875fefa2ed41dd2d29b44bccc807ec6f8bc9fb74e3a2a6a6f50f236d99e  guix-build-28725337e601/output/arm-linux-gnueabihf/bitcoin-28725337e601-arm-linux-gnueabihf.tar.gz
5e910973159116c6beb5f1dbb86990baaa6a6b8da8e3f621462b17d3ab1dcdc9  guix-build-28725337e601/output/dist-archive/bitcoin-28725337e601.tar.gz
fdb11bd1f5a99f4fdadd35ce34f84884a2c31be030ca3c441d82feb1ec277c8b  guix-build-28725337e601/output/powerpc64-linux-gnu/SHA256SUMS.part
2121b104948e0f4a9677ccd58dcb05cc4f4f9adbd080fea8072714899763b639  guix-build-28725337e601/output/powerpc64-linux-gnu/bitcoin-28725337e601-powerpc64-linux-gnu-debug.tar.gz
8bf13d1a32bc15b2e92b86103c63fb02ad7c65b69065b8ce3963756eda72ce6d  guix-build-28725337e601/output/powerpc64-linux-gnu/bitcoin-28725337e601-powerpc64-linux-gnu.tar.gz
18910b8da72c9aab0a5642923826b7dba6c89a5f1d60aa84524a5f38e118bdd1  guix-build-28725337e601/output/powerpc64le-linux-gnu/SHA256SUMS.part
289be1f144ccfa1d6e4f65a835d1bc759bd5e8b0c47d8a38e970fbccfd8a7e48  guix-build-28725337e601/output/powerpc64le-linux-gnu/bitcoin-28725337e601-powerpc64le-linux-gnu-debug.tar.gz
ae8055a3564fab10f88de7f7b51a0317da1424e6b52d0d6ac43ca7ddbfbfceb5  guix-build-28725337e601/output/powerpc64le-linux-gnu/bitcoin-28725337e601-powerpc64le-linux-gnu.tar.gz
ef5e268e26d50776f815a8d915cafb0318947eb14880ca5d17d6e1d2fd820216  guix-build-28725337e601/output/riscv64-linux-gnu/SHA256SUMS.part
dbd0bdc40fd1ce847d9d1d8411303b37b914c10837d3308186f86facc6f0bd3b  guix-build-28725337e601/output/riscv64-linux-gnu/bitcoin-28725337e601-riscv64-linux-gnu-debug.tar.gz
efa2858f9e04c9575309d9b34fdcd68fa0ea2fbdaad5293754b4d5ac2bc1f25e  guix-build-28725337e601/output/riscv64-linux-gnu/bitcoin-28725337e601-riscv64-linux-gnu.tar.gz
fd6c6de0b9ca56595ce6e511cd69780f76d7b1e7f4c2405e47aa44ee73fdc7f4  guix-build-28725337e601/output/x86_64-apple-darwin/SHA256SUMS.part
17f098049726196847752c6d2bbc1e9c1cf8122f6631b3df52e5588c1895b4fc  guix-build-28725337e601/output/x86_64-apple-darwin/bitcoin-28725337e601-osx-unsigned.dmg
0740bfb2753e7a8aad4c953fb638be4f96db9194dac94158ccd08a657a22099b  guix-build-28725337e601/output/x86_64-apple-darwin/bitcoin-28725337e601-osx-unsigned.tar.gz
95dd7a6370486906c26564eeafd1ae08ce05e287a1ec86d5e3c61a392257fce3  guix-build-28725337e601/output/x86_64-apple-darwin/bitcoin-28725337e601-osx64.tar.gz
f3516a46d7019f8b473901b0599af3f494f0a1b487a33ee21d4fd1742767894d  guix-build-28725337e601/output/x86_64-linux-gnu/SHA256SUMS.part
8c30a6e0fa7317ff5e84def04a39ac09e062e19010c613c17c05ef1cf1312660  guix-build-28725337e601/output/x86_64-linux-gnu/bitcoin-28725337e601-x86_64-linux-gnu-debug.tar.gz
ed9c1d4872478ac59a3f49ebe2640d51cf1882335bb94aac5261f15e3f2e3a27  guix-build-28725337e601/output/x86_64-linux-gnu/bitcoin-28725337e601-x86_64-linux-gnu.tar.gz
4d4c2adae87d52b99880ff9b1adbf08abddef7060432e1da800278e009c1cd13  guix-build-28725337e601/output/x86_64-w64-mingw32/SHA256SUMS.part
de6268e155e04f73226f9513dd3a44322c96abdf06792538bccf2f7a7d1f453b  guix-build-28725337e601/output/x86_64-w64-mingw32/bitcoin-28725337e601-win-unsigned.tar.gz
e2732734e018100e513653bc22f113402a8a5b0aea79c0f39728e8ff4a234152  guix-build-28725337e601/output/x86_64-w64-mingw32/bitcoin-28725337e601-win64-debug.zip
008755bbb1ec6076b6fd27ebf9fe53edc49458918773429a7a2f440b6ed44cfd  guix-build-28725337e601/output/x86_64-w64-mingw32/bitcoin-28725337e601-win64-setup-unsigned.exe
d2ef854bdb2603a477d0e98049394342966c01713c231e93b55abe3dcac9a503  guix-build-28725337e601/output/x86_64-w64-mingw32/bitcoin-28725337e601-win64.zip
```

dongcarl reviewed (commented):
Looking good so far, would be great if we pass these along to bitcoin core's configure via `config.site`!

hebasto committed:
build: Propagate TOOL_FOR_BUILD variables to the configure script

hebasto commented:
Updated 28725337e6011789a6b4e80c0cb127979bf34221 -> bcfe469576156aaf2f9a33fa8f083910c3266143 ([pr23571.03](https://github.com/hebasto/bitcoin/commits/pr23571.03) -> [pr23571.04](https://github.com/hebasto/bitcoin/commits/pr23571.04)).

Addressed the recent @dongcarl's comments.

hebasto commented on depends/builders/default.mk:
```diff
@@ -8,11 +8,17 @@ default_build_OTOOL = otool
 build_$(build_os)_$1 ?= $$(default_build_$1)
 build_$(build_arch)_$(build_os)_$1 ?= $$(build_$(build_os)_$1)
-build_$1=$$(build_$(build_arch)_$(build_os)_$1)
+endif
```
Thanks!

[Updated](https://github.com/bitcoin/bitcoin/pull/23571#issuecomment-999116510).

dongcarl commented:
Code Review ACK bcfe469576156aaf2f9a33fa8f083910c3266143 🚀

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
5c4db77a31bd8b696fae815a357ce06152a75cea4706f0acb36904b4521cc64b  guix-build-bcfe46957615/output/aarch64-linux-gnu/SHA256SUMS.part
a835c59f6aa02754a2bed6ad6b995adef9b672f6bf76806a09dcd45fecdf28a9  guix-build-bcfe46957615/output/aarch64-linux-gnu/bitcoin-bcfe46957615-aarch64-linux-gnu-debug.tar.gz
949d274c83b09c9eced8a0f45561fb126a1a3a4fb23163d3080193707c7164ee  guix-build-bcfe46957615/output/aarch64-linux-gnu/bitcoin-bcfe46957615-aarch64-linux-gnu.tar.gz
902ddca13a5358271376e23f1d65adad1b2b8f0aa775be0ab17196b78f050ed7  guix-build-bcfe46957615/output/arm-linux-gnueabihf/SHA256SUMS.part
6bcf0380c54dd0f3c7742a5965ccdfcf85a0a3a2c3e5aa39fafd24a9b3cdec8b  guix-build-bcfe46957615/output/arm-linux-gnueabihf/bitcoin-bcfe46957615-arm-linux-gnueabihf-debug.tar.gz
d22ee005fea1912a8fd5db7e6bf2d5f0a9a57fee7ef50355b580e167364dcdb4  guix-build-bcfe46957615/output/arm-linux-gnueabihf/bitcoin-bcfe46957615-arm-linux-gnueabihf.tar.gz
842134d949c14e178376472b1a607fadf01d54969871804557bb356469bf527f  guix-build-bcfe46957615/output/dist-archive/bitcoin-bcfe46957615.tar.gz
aa3531ed9347a7b3153bd42b89a079ee5b58655f87bbc8eb379ae94fd4eb76b4  guix-build-bcfe46957615/output/powerpc64-linux-gnu/SHA256SUMS.part
99836c8e17fd3c82b9b66d47d43aef21a03b0cbac753abc1d05467de463fa89d  guix-build-bcfe46957615/output/powerpc64-linux-gnu/bitcoin-bcfe46957615-powerpc64-linux-gnu-debug.tar.gz
f9348f571e58a15bfcdb5a75f8b161f0e9b100a785a3b9a0a824b076da24f363  guix-build-bcfe46957615/output/powerpc64-linux-gnu/bitcoin-bcfe46957615-powerpc64-linux-gnu.tar.gz
3095f92ac9cd336edd4468386a11ee7b3281c9aef699475f11c5999cce74f1df  guix-build-bcfe46957615/output/powerpc64le-linux-gnu/SHA256SUMS.part
060bd1a58748125b018827d65e2d5062e52f7d641ed88ccede9bb856365c068a  guix-build-bcfe46957615/output/powerpc64le-linux-gnu/bitcoin-bcfe46957615-powerpc64le-linux-gnu-debug.tar.gz
f9de2f5cbe14f4a5f567b590790f6878447a0b7e6fe95ca0d6139ab3de4d8bac  guix-build-bcfe46957615/output/powerpc64le-linux-gnu/bitcoin-bcfe46957615-powerpc64le-linux-gnu.tar.gz
8d5234a947ff9350f8b375b4d13c5a54f8b7060301db359c828c4620135a98cb  guix-build-bcfe46957615/output/riscv64-linux-gnu/SHA256SUMS.part
30e0b5d614c90c95fa5c3280c16009535dd77e37c82aca489d454e87957e43e5  guix-build-bcfe46957615/output/riscv64-linux-gnu/bitcoin-bcfe46957615-riscv64-linux-gnu-debug.tar.gz
4ceb917473ace0b05d582f269c1801c026532310a22ef3cedf299cd42613ff77  guix-build-bcfe46957615/output/riscv64-linux-gnu/bitcoin-bcfe46957615-riscv64-linux-gnu.tar.gz
24de692eeb2fcad2938339fc2dbeca08f1763c3aab15b4c2c3791fd96a6055f7  guix-build-bcfe46957615/output/x86_64-apple-darwin/SHA256SUMS.part
dd073eddaeaa98e279758b8524e71ac96fab1f6805b7f3a96297189dfb8b06bf  guix-build-bcfe46957615/output/x86_64-apple-darwin/bitcoin-bcfe46957615-osx-unsigned.dmg
db5e6d0106a0a61594d82f2d9411bbd905a0b36af5c425bc147ba174c7ad264a  guix-build-bcfe46957615/output/x86_64-apple-darwin/bitcoin-bcfe46957615-osx-unsigned.tar.gz
304dc2bbe88fee8449b65552857adaa3ebeb3fecd975dbb077384fad20cd0ad5  guix-build-bcfe46957615/output/x86_64-apple-darwin/bitcoin-bcfe46957615-osx64.tar.gz
3d7c67322e1135d033d8bd811fa33df1c52a53c88ccd81e060efb5eb7915f959  guix-build-bcfe46957615/output/x86_64-linux-gnu/SHA256SUMS.part
ccae1dd7d6b4edaa45c999340b50031e5b88259b16e76edf49721cc75a19dc07  guix-build-bcfe46957615/output/x86_64-linux-gnu/bitcoin-bcfe46957615-x86_64-linux-gnu-debug.tar.gz
175a70d6275b6c108a6bcc4623c6af35cf6587fa3afc086e141e3b7b5c09f989  guix-build-bcfe46957615/output/x86_64-linux-gnu/bitcoin-bcfe46957615-x86_64-linux-gnu.tar.gz
8bba3659687987954588f9d3dd6899080db92e2b4e694772724abb3e13b0141c  guix-build-bcfe46957615/output/x86_64-w64-mingw32/SHA256SUMS.part
43ae47c0479aa0de6062a61daae1f06fa629e91e362caf5458113d85de6dccb2  guix-build-bcfe46957615/output/x86_64-w64-mingw32/bitcoin-bcfe46957615-win-unsigned.tar.gz
5f9b865cb5875d9628ec085a5b66af859c88db651dadc325ac385b1a733c76ce  guix-build-bcfe46957615/output/x86_64-w64-mingw32/bitcoin-bcfe46957615-win64-debug.zip
e3629bf7e9e975566f780d533ada4d0a9557fa541d53dcd4e4d044173d377642  guix-build-bcfe46957615/output/x86_64-w64-mingw32/bitcoin-bcfe46957615-win64-setup-unsigned.exe
ff1dda1c43449db52bcd93525e59b5459897c5742d030314110f4efa052a69c7  guix-build-bcfe46957615/output/x86_64-w64-mingw32/bitcoin-bcfe46957615-win64.zip
```

hebasto commented:
Rebased bcfe469576156aaf2f9a33fa8f083910c3266143 -> 5f7b00416e291f11b93222c896e4963fc50528a9 ([pr23571.04](https://github.com/hebasto/bitcoin/commits/pr23571.04) -> [pr23571.05](https://github.com/hebasto/bitcoin/commits/pr23571.05)).

laanwj commented:
Light code-review ACK 5f7b00416e291f11b93222c896e4963fc50528a9

hebasto commented:
Rebased 5f7b00416e291f11b93222c896e4963fc50528a9 -> 76d4d3bb407325bb0350de45ece982e286a1087f ([pr23571.05](https://github.com/hebasto/bitcoin/commits/pr23571.05) -> [pr23571.06](https://github.com/hebasto/bitcoin/commits/pr23571.06)) due to the conflict with #19952.

hebasto commented:
#### Guix builds on `x86_64`:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
d2192fb917e24c840dfac5e85636cbe533bae891cd9affdd9f88b346fc524411  guix-build-76d4d3bb4073/output/aarch64-linux-gnu/SHA256SUMS.part
e2a46151d0a176026d51849ada443baa0d13001621ef353bbfcd2fbca884df6d  guix-build-76d4d3bb4073/output/aarch64-linux-gnu/bitcoin-76d4d3bb4073-aarch64-linux-gnu-debug.tar.gz
a6944abf7590f39b19f23f79f372fed020ed8af39003a956ab205038de75b4a0  guix-build-76d4d3bb4073/output/aarch64-linux-gnu/bitcoin-76d4d3bb4073-aarch64-linux-gnu.tar.gz
1b282042bdbc3d5418f90c6eb26f5ef199b632149ead0688d4b4607a7373f7ea  guix-build-76d4d3bb4073/output/arm-linux-gnueabihf/SHA256SUMS.part
342b92c8366b6d99508f5ebe678870feb3a1b117c485606ca5966c2751c2a588  guix-build-76d4d3bb4073/output/arm-linux-gnueabihf/bitcoin-76d4d3bb4073-arm-linux-gnueabihf-debug.tar.gz
006d00b8ce4b2b3a1ca793976b971274a2ad0af3e126dd7573f8ae6f62fabd70  guix-build-76d4d3bb4073/output/arm-linux-gnueabihf/bitcoin-76d4d3bb4073-arm-linux-gnueabihf.tar.gz
bccd8bb8fdb49808f9170884a637f7f7d32bdc98c8b842994e9df5666fe7213a  guix-build-76d4d3bb4073/output/arm64-apple-darwin/SHA256SUMS.part
c048ee3527c46d85b69ff10e15d92f6900a5763ed85446731306009494d76515  guix-build-76d4d3bb4073/output/arm64-apple-darwin/bitcoin-76d4d3bb4073-arm64-apple-darwin-unsigned.dmg
04ca03d1e9df281bd6b0ce5a581e2628efb40b129b92423f42e3183fc3bad5f9  guix-build-76d4d3bb4073/output/arm64-apple-darwin/bitcoin-76d4d3bb4073-arm64-apple-darwin-unsigned.tar.gz
d9d6f171bae71f429e269525cdceebbafd3caae161a3f996d9b06714229b8c2a  guix-build-76d4d3bb4073/output/arm64-apple-darwin/bitcoin-76d4d3bb4073-arm64-apple-darwin.tar.gz
9e77ef269d8c444e3db88d47e9edc8f1bf07b51de10e7ee50db85b568b421835  guix-build-76d4d3bb4073/output/dist-archive/bitcoin-76d4d3bb4073.tar.gz
734a697b081eacb02af84c306b496145f16a1a2c437dd1cf2c89516546f30e49  guix-build-76d4d3bb4073/output/powerpc64-linux-gnu/SHA256SUMS.part
683696b5199e42239136fdce4b376f7d61908e0333f7179a4e7845870b39200f  guix-build-76d4d3bb4073/output/powerpc64-linux-gnu/bitcoin-76d4d3bb4073-powerpc64-linux-gnu-debug.tar.gz
22ca1f7e04a79a3cb7d938ef7e7a5446c30bca00d07e289fcd5c5674b1cd41c9  guix-build-76d4d3bb4073/output/powerpc64-linux-gnu/bitcoin-76d4d3bb4073-powerpc64-linux-gnu.tar.gz
03b1f5baaf90b9bb8fcf13b2ceecf7aab500087c1e60272be372b04e0f79a597  guix-build-76d4d3bb4073/output/powerpc64le-linux-gnu/SHA256SUMS.part
ae248964c7dd3043ab3f36aa1f1b7012ce19c7b6629616688ca18b62ee738ab2  guix-build-76d4d3bb4073/output/powerpc64le-linux-gnu/bitcoin-76d4d3bb4073-powerpc64le-linux-gnu-debug.tar.gz
aab70af704c73443c2669e44cf9e9ae03a82ce5cc499b1e0fb92ec887b1ef664  guix-build-76d4d3bb4073/output/powerpc64le-linux-gnu/bitcoin-76d4d3bb4073-powerpc64le-linux-gnu.tar.gz
b145ec41f4a93c776c86b315a1d72205fae90ad14b9011db5ed32c08a1f775c8  guix-build-76d4d3bb4073/output/riscv64-linux-gnu/SHA256SUMS.part
3fbeaed5ff8af6c99dfc0d8bdddb9e3e9d10cf9b3213a06ac07aa73ac437979d  guix-build-76d4d3bb4073/output/riscv64-linux-gnu/bitcoin-76d4d3bb4073-riscv64-linux-gnu-debug.tar.gz
cb9e23c2dc5bc9d4b463cf93cb7d458252f7329ba84ca0149f48b9a227e35eb7  guix-build-76d4d3bb4073/output/riscv64-linux-gnu/bitcoin-76d4d3bb4073-riscv64-linux-gnu.tar.gz
945ac368b3ba5199ce886b3d29756612a2178c994aac0f8fc1257d76e4a12178  guix-build-76d4d3bb4073/output/x86_64-apple-darwin/SHA256SUMS.part
bd958cbae71999233e75e84ed5245e55058bb9477aa984081dc188c2aeb7baa3  guix-build-76d4d3bb4073/output/x86_64-apple-darwin/bitcoin-76d4d3bb4073-x86_64-apple-darwin-unsigned.dmg
22823fe2571ba1ed96337d07929b239eec9dbca2c2bc57ee547477426f0ab573  guix-build-76d4d3bb4073/output/x86_64-apple-darwin/bitcoin-76d4d3bb4073-x86_64-apple-darwin-unsigned.tar.gz
40950a2635f94ba12aab03a15044ff6adcac5dac62328c6f650e1b0a48d2bfd3  guix-build-76d4d3bb4073/output/x86_64-apple-darwin/bitcoin-76d4d3bb4073-x86_64-apple-darwin.tar.gz
6845e0499d8eb600e21e5676e833bd6fa3fdb1f40b6bdaf23d90ccd247cccd54  guix-build-76d4d3bb4073/output/x86_64-linux-gnu/SHA256SUMS.part
dc874a0c577aaf1da46f6f71820352dcb7bb6aadceed68f8d1f77c00a2da72c5  guix-build-76d4d3bb4073/output/x86_64-linux-gnu/bitcoin-76d4d3bb4073-x86_64-linux-gnu-debug.tar.gz
2367af4b41e3c031ed8e14d741ce046d454f9973d99f829eed6811c47937d695  guix-build-76d4d3bb4073/output/x86_64-linux-gnu/bitcoin-76d4d3bb4073-x86_64-linux-gnu.tar.gz
fc8e97f171049736b0e4a5e5a854dc9442de3ba4f640981b0eccb5f29734d034  guix-build-76d4d3bb4073/output/x86_64-w64-mingw32/SHA256SUMS.part
28577d84cad6404cfc5224b68285496c7be975458156f14f7d971dc3cb6dabf4  guix-build-76d4d3bb4073/output/x86_64-w64-mingw32/bitcoin-76d4d3bb4073-win64-debug.zip
8f0bf22fec30f8eeeb702d451a06d478eddf45724126e9c9f3741587e31ca0cd  guix-build-76d4d3bb4073/output/x86_64-w64-mingw32/bitcoin-76d4d3bb4073-win64-setup-unsigned.exe
4c891a7a4a679a02a208b5e8f31f2ac319e8c109501165e5025e3c988cf3e5cd  guix-build-76d4d3bb4073/output/x86_64-w64-mingw32/bitcoin-76d4d3bb4073-win64-unsigned.tar.gz
f511049e13c44282d4d744d0dba996eb051b67d79d624fea81b68e6a3c9fd820  guix-build-76d4d3bb4073/output/x86_64-w64-mingw32/bitcoin-76d4d3bb4073-win64.zip
```

hebasto commented:
Rebased 76d4d3bb407325bb0350de45ece982e286a1087f -> ef802d0a714b0d6d149c950e323c56d82916d690 ([pr23571.06](https://github.com/hebasto/bitcoin/commits/pr23571.06) -> [pr23571.07](https://github.com/hebasto/bitcoin/commits/pr23571.07)).

achow101 commented:
Are you still working on this?

sedited reviewed (approved):
ACK ef802d0a714b0d6d149c950e323c56d82916d690

fanquake commented:
Can probably be rebased for #23619. I'm not super convinced at adding all the `_FOR_BUILD` infra given there is currently little use for it outside of a single optional experimental package.

hebasto commented:
@fanquake 
> Can probably be rebased for #23619.

Done.

> I'm not super convinced at adding all the `_FOR_BUILD` infra given there is currently little use for it outside of a single optional experimental package.

Are you suggesting to drop last two commits or to close this PR altogether?

fanquake commented:
> Are you suggesting to drop last two commits

I think it would be better if this was scoped to just the actual bug fix.

hebasto commented:
Closing. Up for grabs.
